R/DataMart.R

```diff
@@ -483,3 +483,221 @@ IsDisplayrCloudDriveAvailable <- function()
     company.secret <- get0("companySecret")
     !is.null(company.secret) && company.secret != "UNKNOWN"
 }
+
+uploadRScript <- function(r.code,
+                          js.code = NULL,
+                          filename,
+                          upload = TRUE,
+                          api.root = Sys.getenv("API_ROOT"),
+                          company.secret = Sys.getenv("COMPANY_SECRET"),
+                          client.id = Sys.getenv("CLIENT_ID"))
+{
+    if (missing(filename))
+        stop(substitute(filename), " argument required as ",
+             "filename to write in Displayr Drive is required.")
+    type <- file_ext(filename)
+    stopifnot("rscript filename extension required" = type == "rscript")
+    if (upload)
+        checkUploadPossible(api.root = api.root, company.secret = company.secret, client.id = client.id)
+    tmpfile <- tempfile()
+    file <- file(tmpfile, "wb")
+    r.filenames <- if (is.list(r.code)) unlist(list(r.code)) else r.code
+    for (current.file in r.filenames)
+    {
+        cat("# R code below\n", file = tmpfile, append = TRUE)
+        file.append(tmpfile, current.file)
+    }
+    if (!is.null(js.code))
+    {
+        cat("\n\n// JS Code below\n\n", file = tmpfile, append = TRUE)
+        js.filenames <- if (is.list(js.code)) unlist(list(js.code)) else js.code
+        for (current.file in js.filenames)
+            file.append(tmpfile, current.file)
+    }
+    close(file)
+    if (upload)
+    {
+        res <- try(POST(paste0(api.root, "?filename=", URLencode(filename, TRUE)),
+                        config = add_headers("Content-Type" = guess_type(filename),
+                                             "X-Q-Company-Secret" = company.secret,
+                                             "X-Q-Project-ID" = client.id),
+                        encode = "raw",
+                        body = upload_file(tmpfile)))
+        if (!inherits(res, "try-error") && res$status_code == 413)  # 413 comes from IIS when we violate its web.config limits
+            stopBadRequest(res, "Could not write to Displayr Cloud Drive. Data to write is too large.")
+        else if (inherits(res, "try-error") || res$status_code != 200)
+        {
+            warning("uploadRScript has encountered an unknown error.")
+            stopBadRequest(res, "Could not save file.")
+        }
+
+        msg <- paste("RScript uploaded to Displayr Cloud Drive with name", filename,
+                     sep = "\n")
+    } else
+    {
+        file.copy(tmpfile, filename, overwrite = TRUE)
+        msg <- paste("RScript created locally with name", filename,
+                     "In the directory",
+                     getwd(),
+                     sep = "\n")
+    }
+
+    message(msg)
+    invisible()
+}
+
+uploadStandardR <- function(directory = ".", filename = NULL, upload = TRUE)
```

Comment:
This function currently doesn't work if there are more than one Standard R page in the directory, e.g. `Q-Wiki-Scripts/Standard R/Tables/`.

Reply:
Good point. I had my blinkers on only thinking about `Calculation` where they are all in a separate directory. Perhaps the API should check, if a directory is provided then a single JS and R file is required. Otherwise the filename within the directory should be required. I'll make a change to allow this.

Reply:
Fixed in https://github.com/Displayr/flipAPI/pull/12/commits/59ec801d6a5bf16535bb37519f10841a1b690a32